### PR TITLE
remove compatibility for API levels < 15

### PR DESCRIPTION
**Pull request by mikehardy**
I grep'ed through the source code for version dependencies and cleaned out <= 14

Note that api 14 was never in wide release and doesn't even show up on the dashboards since google pushed manufacturers to 15 quickly. If we went to 15, we could get ride of a bit more

Also, we may be able to get rid of the CompatVNN classes as well, but I didn't want to even attempt that without discussion. If that is desired I can do that.

Finally, there is a change to the android ImageView compat class as well - this was suggested by AndroidStudio but I could unwind that easily if it is not desired

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java [hunk 1]
```diff
@@ -367,20 +364,18 @@ public void onClick(View v) {
...
+                            mRecorder = initMediaRecorder();
+                            mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
+                            mRecorder.setAudioChannels(2);
+                            mRecorder.setAudioSamplingRate(44100);
+                            mRecorder.setAudioEncodingBitRate(192000);
+                            mRecorder.prepare();
+                            mRecorder.start();
+                            highSampling = true;
+                        } catch (Exception e) {
+                            // fall back to low sampling
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: Where is the exception thrown?

**Comment by timrae** [comment]
Please do also remove the CompatV11,12,14.
Also I'm ok to go straight to v15 if it's useful in anyway.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: Hmm - I'll check - previously it was just an empty block, and my change was just to add a comment, but I'll inspect. The whole thing (high-sample try, fallback to low) was guarded by an API level check and on this one I conservatively left the fallback just in case but maybe high sample rate is always available and if there is an exception it's real and should cause logging + error messaging?

**Comment by mikehardy** [comment]
Okay, your public announcement on the mailing list didn't mention API levels, it just mentioned "KitKat" so I don't think this is even a change of messaging. API 14 was buried pretty quickly and API 15 would allow a bit more cleanup. I'll look into that as well as removal of the compat classes. A good intro task to how all the compat stuff works, as a useful side track while I delay looking at the Oreo notification channels will likely need compat work

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java (hunk 1 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: Yeah it could well be, only the documentation will tell. In general I'd rather target the exact exception if it's real.

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: Agreed - if catching Exception why not just go for Throwable or Error? (no, please no...)

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: It throws IOException on MediaRecorder.prepare() if anything relating to setting up the output file fails (fair enough). 

.prepare() can also throw unchecked IllegalArgumentException as can many of the other codec setup calls

I'm guessing that's why the original code just caught Exception, because the handling case is the same - for any error try low fidelity, and if errors continue, toast and stop

I'm not sure this can be improved then other than commenting the code section clearly

**Comment by mikehardy** [comment]
I attempted to install and run an API 15 emulator to check things and Anki won't even run because it can't create the storage directories on the emulator, so this will sit for a moment while I figure that out

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java (hunk 1 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/AudioView.java]: Ok thanks for checking, a comment would definitely be helpful!

**Comment by timrae** [comment]
Hmmm... do you get an exception?

**Comment by mikehardy** [comment]
I'm most of the way done moving AnkiDroid to API >= 15, but what about libanki?  

I have one remaining dependency on "isHoneycomb()" and it's in DB.java which I think is in libanki which still has a minSDK of 8

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 2]
```diff
@@ -724,19 +690,12 @@ private void removeUnnecessaryAdvancedPrefs(PreferenceScreen screen) {
...
-            }
-            if (CompatHelper.getSdkVersion() >= 15) {
-                workarounds.removePreference(writeAnswersDisable);
-                workarounds.removePreference(inputWorkaround);
-            } else {
-                // For older Androids we never use the input tag anyway.
-                workarounds.removePreference(useInputTag);
-            }
+
+            workarounds.removePreference(longclickWorkaround);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: These workarounds should be able to be removed entirely I believe...

Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: For example the long click workaround:

https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java#L1340

**Comment by timrae** [comment]
Sorry for the laaate reply on this. I added a couple of comments, otherwise it looks it good, thanks.
As for the [`DB.java`](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/DB.java#L61) checks, you can safely remove those as well. Libanki is just a folder of source code in the main project, as such it has the same SDK requirements as the other AnkiDroid code. Maybe you were confused with the AnkiDroid API.

**Comment by mikehardy** [comment]
Great - thanks for the review and I'll look more closely into the workarounds you commented on. I'm naive to their effects so didn't feel confident about them obviously! Also thanks for the clarification on DB.java, I was confused in exactly the way you mention - unsure if they were involved in the API or not.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java (hunk 2 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: Latest commit removes all the workarounds including all the logic+strings+css etc that spidered out from it. The only thing I didn't remove was the "safeDisplayMode" - turns out this was a combo of API < 15 and device == Nook, I made the check for this preference Nook specific, and updated the related comments for the safeUpdate logic

**Comment by mikehardy** [comment]
Okay, this is ready for review again

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 3]
```diff
@@ -716,27 +682,13 @@ private void removeUnnecessaryAdvancedPrefs(PreferenceScreen screen) {
...
+
+            if (!CompatHelper.isNook()) {
                 workarounds.removePreference(safeDisplayMode);
             }
-            if (CompatHelper.getSdkVersion() >= 15) {
-                workarounds.removePreference(writeAnswersDisable);
-                workarounds.removePreference(inputWorkaround);
-            } else {
-                // For older Androids we never use the input tag anyway.
-                workarounds.removePreference(useInputTag);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: This setting isn't used anymore right? I think you can remove it from the Reviewer code as well...?

Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: Oh sorry I just understood lol

File: AnkiDroid/src/main/java/com/ichi2/libanki/DB.java [hunk 4]
```diff
@@ -57,29 +55,19 @@
...
-                            | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
-        }
 
-        if (mDatabase != null) {
-            // TODO: we can remove this eventually once everyone has stopped using old AnkiDroid clients with WAL
-            CompatHelper.getCompat().disableDatabaseWriteAheadLogging(mDatabase);
-            mDatabase.rawQuery("PRAGMA synchronous = 2", null);
-        }
-        // getDatabase().beginTransactionNonExclusive();
+        // TODO: remove this once everyone has stopped using old AnkiDroid clients with WAL (API >= 16)
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: Why API >= 16? It's related to the AnkiDroid version rather than the API version...

Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: Is removing the null check safe?

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: A good question on API >= 16, I was just re-homing the comment from CompatHelper, intended as a breadcrumb for when this is okay to delete? See here: https://github.com/ankidroid/Anki-Android/pull/4793/files/efa3e8ddc57d9902aa0c6bd97b2a1a716467fd23#diff-77adff8eacc9a02363d6bd3e31d33d02L36

You'll know better on that one - I was just keying off the honeycomb note is all

As for the null check, the openDatabase implementation says that it either returns the database or throws SQLiteException, so it *should* be safe. I don't like the word "should" but it looks like it has been checked and not checked in the past (using the blame function) but never in reference to a logged issue, it was just put in there during the journal mode changes 7 years ago and has sat there since? https://github.com/ankidroid/Anki-Android/blame/d661a260884dd50e65f3b3df63597d9627e40cce/src/com/ichi2/anki/AnkiDb.java#L46

Can't really hurt to do a null check except maybe performance and this is not a performance-sensitive area so happy to leave it just in case

**Comment by timrae** [comment]
Merge commits inside pull requests kind of make the commit history overly complicated, so we usually ask contributors to rebase instead. Do you know how to get rid of the merge commit? The easiest way IMO is to re-pull the master branch, cherry-pick your commits onto that, and force push with the same branch name.

**Comment by mikehardy** [comment]
git merge/push stuff is the area of git that is admittedly foreign to me but it is also one of the reasons I'm doing this - to learn. I've done something similar on this project once already and I know there are guides for all of it all over the net, so I'll make it happen, sure

File: AnkiDroid/src/main/java/com/ichi2/libanki/DB.java (hunk 4 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: > I was just re-homing the comment from CompatHelper, intended as a breadcrumb for when this is okay to delete

IIRC the comment is actually saying that we can remove the whole process of disabling of WAL once there are no clients in circulation that still use WAL. It's not related to the issue of whether or not the framework provides a native method to disable it or not, rather it was just added as an extra safety measure in #3856. That was over 2 years ago now, so I think we can safely remove it now even.

Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: > As for the null check, the openDatabase implementation says that it either returns the database or throws SQLiteException, so it should be safe.

Yeah that seems to be correct. What's this `SQLException` in the function definition though? Should we in-fact be catching the `SQLiteException`

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/DB.java]: Okay - for the WAL changes I'll purge the last bit there.

For the SQLiteException, it extends RuntimeException so we could leave it. Usage of the DB class constructor is mostly (!) in try/catch but not all, so I'd assert that calling code should be hardened but that seems like a tangent here - I'd log separate issue and pursue separately. For this block, it appears there's no resource cleanup necessary on exception and the exception was bubbling up previously, so I think the only useful action is log and rethrow. Since the calling code isn't 100% on handling this, and knowing how difficult good user support is, I like the idea of logging anything we can then re-throwing. I'll throw in a try/catch+log+rethrow unless there's an objection

**Comment by timrae** [comment]
@marcardar 
We make a point to support older APIs so that users from developing countries with recycled handsets will have access. We did once consider switching over to multiple APKs on the play store, but as I recall there were some disadvantages or difficulties for us there, and we decided it wasn't worth it at the time. Things may have changed since then with the play store, so it could be worth revisiting that decision (given infinite time :-p)

**Comment by marcardar** [comment]
The multiple apk support is much improved compared to a few years. It's super smooth now. Regarding older SDK levels, those users still use the old battle-tested version.  They are probably happy with what they have. Things can break in one of two ways: 1. System update, 2. App update
Fortunately, older SDK levels are not receiving system updates so the only issue is with app updates. You just have to weigh up the cost of testing on older devices (of which there are over 10,000 different models) vs. giving such users the latest features. I suspect most of them would rather not take the risk of breaking what they already have. If time is an issue, for sure increasing the minSdkVersion is MUCH less work than maintaining the latest features on older devices.

**Comment by mikehardy** [comment]
Wow, that took me a while to get to. That should be two lessons for me: 1) don't commit a merge on a branch, and 2) keep commits very granular since the conflict was in a commit jumbled up with all sorts of other changes. Not a problem really but needlessly laborious.

Pending a re-review, if it was good before the merge commit it should also be good now - my diffs locally look clean again, but without the merge

I'll be glad to have this one off my workbench...

**Comment by mikehardy** [comment]
Separately - with regard to multiple APKs that's an interesting idea, but the compat infrastructure in AnkiDroid (which I was just all over...) is pretty good in my opinion? And right now I don't think the workarounds in place are actually that heavy though I could of course be missing something obvious. Compared with a "what version are you on" conversation during every support instance it might be about the same support wise. We don't really do a lot of version-testing anyway do we? Just sort of fix things as users tell us they're broken is what I've seen in the last few months anyway.

That said, when we do move forward and clip off old SDKs seems like there's no reason not to leave them with the last valid version, and then maybe if there was some way to keep track of which SDK is on which version tag in case there was some sort of easy showstopper bug to backport, it would be possible to even offer some minimal ongoing support.

**Comment by marcardar** [comment]
Hi Mike - regarding "what version you are on?" - well, you have to do this anyway because you are increasing the minSdkVersion, and even if you weren't, it is always possible someone is on an old version (even if they are running Android P). In my experience, the vast majority (>95%) of bugs occur as a result of app (or system) updates (AnkiDroid's MMV I suppose). Battle-tested code rarely throws up surprises.  It might be worth going over past AnkiDroid Issues and looking at how many of those were introduced in an app update and only occurred on older SDK levels.  Just a quick glance at recent issues - I see [this](https://github.com/ankidroid/Anki-Android/issues/4525#issuecomment-367820188) type of logic which is necessary but not much fun for a developer.

FWIW, my minSdkVersion timeline looks like this:

October 2009 (first release): 3 (Cupcake 1.5 was released April 2009)
August 2012: 9 (Gingerbread 2.3 - December 2010)
January 2015: 16 (JellyBean 4.1 - July 2012)
March 2018: 23 (Marshmallow 6.0 - October 2015)

In other words, about a 2.5 year support level (and if the device manufacturer provides 2 years support then this, in effect, is 4.5 years). Note, the [Nexus 4](https://en.wikipedia.org/wiki/Nexus_4) (stuck on Lollipop) was released in November 2012 (Jelly Bean 4.2) - which works out at over 5 years support (when increasing minSdkVersion to 23).

**Comment by mikehardy** [comment]
I agree with the idea of battle-tested code being worth it's weight in gold, and I like the idea of classification of issues by cause - e.g., fundamental error, app-update-introduced, system-update-introduced but I wouldn't be able to go back in time to do it. I can say most of the issues I'm working on now are system-update-introduced in that AnkiDroid doesn't use APIs correctly given new restrictions on existing APIs but I can also say even with the linked commit (setInexactRepeating vs setRepeating) this hasn't been something that's hard to make work on all prior versions, the real work (for me) is in understanding the API itself. That's not an argument against different APKs though. Are there good documents around on exactly how to do a release where you limit the APK to a minimum version, and also how to do essentially the opposite - how to release a fix for a fundamental error to the APK for older versions? There is release procedure documentation for AnkiDroid and assuming any switch to this style there'd need to be a pull request for the docs that explained in detail how to do everything...

**Comment by marcardar** [comment]
It's interesting that most of the issues are system-update-introduced. But anyway, how often are those issue for system updates pre-sdk23?  I suspect very rare, in which case, this also doesn't stengthen the case for keeping the minSdk level low.

Regarding documentation of how to update a lower SDK apk - now that you've raised the minSdkVersion, you have this problem anyway.  In terms of releasing multi-apk on Google Play, this is trivial. In terms of informing contributors how to make a pull request for a prior APK, I can't imagine that will happen often (if ever) since there will be likely no app/system updates to cause them.

All I'm saying here is that now you've made the decision to increase the minSdkVersion, it might be worth increasing it higher still.  There's something very satisfying about stripping out old-sdk-specific code. It makes the code smaller and more readable. Also something very reassuring about not touching battle-tested code!

**Comment by marcardar** [comment]
Actually, when you say system update, do you mean things like a Chrome update affecting the WebView rendering?  Chrome became in the [default renderer](https://developer.chrome.com/multidevice/webview/overview) in Kitkat, so (if true that WebView functionality can change) maybe that would be an understandable reason to make the minSdkVersion kitkat (or lower).

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java (hunk 2 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: @mikehardy 
I think Nook only supports Android 2.2, so you can still remove that

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java]: Ah - I'll confirm that and then nuke the nook support (including safeDisplayMode) if so. Thanks for the review

**Comment by timrae** [comment]
@mikehardy 
Thanks a lot for taking care of this, the "nook" related workarounds can all be removed, as I believe those things only ever worked on Android 2.2. If you could submit a new PR for that when you get time, it'd be appreciated!

**Comment by mikehardy** [comment]
You're welcome! I'm excited to have a big merge go in successfully honestly, mobile development is new for me as I said. My hope is that this one (with all the compatibility work involved) informs my goal of forward compatibility for 7+, fingers crossed. I'll submit a new PR for the nook stuff when I can.

**Comment by timrae** [comment]
@marcardar 
Yeah I need to take a look at the play store and see what the process is like for multiple APKs in 2018, but if there are no significant disadvantages then I'd definitely be onboard with just leaving the "battle tested" 2.8.x available for devices older than Kit-kat and setting that as the new min SDK going forward. With some (substantial) work, this could finally allow us to get rid of that nasty storage permission which would be a huge win.

I think we'd better keep supporting Kit-Kat for a while longer, as that still has a 10% market share, and we have Anki 2.1 (and more?) coming soon, which is going to require some changes that I don't think those users should be excluded from.
